Added edd_profile_editor_before_name action to profile editor

rubengc opened the pull request:
#5148

pippinsplugins commented on templates/shortcode-profile-editor.php:
```diff
@@ -28,6 +28,7 @@
 
 	<form id="edd_profile_editor_form" class="edd_form" action="<?php echo edd_get_current_page_url(); ?>" method="post">
 		<fieldset>
+			<?php do_action( 'edd_profile_editor_before_name' ); ?>
```
Please move this below the leged

rubengc replied:
Ok, but I want to add the section "Change your avatar" before "Change your name"

Would I add a new action for this? (Really I do not want to add an excesive number of actions)

and I just saw that the form does not have the attribute enctype, any idea on how to implement this only when necessary?

SDavisMedia commented:
@spencerfinnell @pippinsplugins Why does this need to be below the legend? Everywhere else on this form, you have the ability to add a completely new field/output unrelated to the existing fields (besides the naming scheme). If you place this hook below the legend, you tie it to the name fields and there's still no way to place something _unrelated_ above the name fields.

@rubengc's Request here is to to start the form with his own output. Placing a hook below the name legend is fine but that's a different goal and it would be incomplete as no other legend in this form are followed immediately by a hook.

spencerfinnell commented:
@SDavisMedia There is no need for it to be above. A legend will always appear at the top of the fieldset, and there should only be one legend per fieldset. 

A separate hook should be added to add the ability to create a separate fieldset/form if that is what is needed.

The only reason this works now is because of the `span` around the legend which should be removed as well. https://github.com/easydigitaldownloads/easy-digital-downloads/issues/3568

http://codepen.io/anon/pen/ObVaYX

spencerfinnell commented:
Might make more sense to submit it against https://github.com/easydigitaldownloads/easy-digital-downloads/issues/3904 as well.

SDavisMedia commented:
@spencerfinnell Good point.

@rubengc Is there a reason why you can't just use `edd_profile_editor_before`? https://github.com/easydigitaldownloads/easy-digital-downloads/blob/master/templates/shortcode-profile-editor.php#L27

rubengc commented:
I added a PR that could be a good point between I need and others could need

First I move edd_profile_editor_before and edd_profile_editor_after inside form
tag to allow add more fieldsets at the start and the end of the form

And I added an action inside form tag to allow add more attributes (like enctype, needle for file uploads)

With this changes we could remove edd_profile_editor_before_name because I think anyone will need more than first, last and display name here

What do you think?

SDavisMedia commented:
@rubengc That's not going to work. You can't change the location of hooks like that because the hooks are already in use. That change will affect sites in an undesirable way.

I see absolutely nothing wrong with the location (line number) you're asking for, but `edd_profile_editor_before` and `edd_profile_editor_after` need to stay exactly where they are.

Instead, create two new hooks located directly above the `<fieldset>` and directly after. Name them `edd_profile_editor_top` and  `edd_profile_editor_bottom`. That way there are hooks before and after the entire form as well as before and after the existing fields in the form.

rubengc commented:
Do you like last PR @SDavisMedia ?

SDavisMedia commented:
@rubengc That looks good to me. And it should behave just fine when #3904 is implemented.

@spencerfinnell See any issues with this approach?

rubengc commented:
@SDavisMedia I see [shortcode-register template](https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/fbe20f79334ef6bd21f6a9810e8e5977bef116df/templates/shortcode-register.php)

Could be a better way use edd_profile_editor_fields_top and edd_profile_editor_fields_bottom to maintain same stucture as other shortcodes as points @spencerfinnell right?

SDavisMedia commented:
@rubengc That works. :+1:

spencerfinnell commented:
Seems like a good approach to me.

rubengc commented:
So now is ready to merge?

SDavisMedia commented:
@rubengc Yup, looking good here. Thanks! I'll let someone finalize it. Thumbs up from me. :+1:

cklosowski commented on templates/shortcode-profile-editor.php:
```diff
@@ -26,7 +26,10 @@
 	<?php do_action( 'edd_profile_editor_before' ); ?>
 
-	<form id="edd_profile_editor_form" class="edd_form" action="<?php echo edd_get_current_page_url(); ?>" method="post">
+	<form id="edd_profile_editor_form" class="edd_form" action="<?php echo edd_get_current_page_url(); ?>" method="post" <?php do_action( 'edd_profile_editor_form_tag' ); ?>>
```
@SDavisMedia @spencerfinnell @rubengc I find it odd to have an action within an HTML tag. The only way this would be useful is to echo out information into the tag, but that's kind of not an ideal situation as there is no guarentee that the person using the hook is going to properly escape, or do proper sanitization of the form field tag.

I'm not sure I see the point of allowing someone to openly run an action on the attributes of a form field, unless I'm missing something.

rubengc commented:
I saw it before:
"And I added an action inside form tag to allow add more attributes (like enctype, needle for file uploads)"

I see this practice on other plugins (not from edd team)

There is another possibility to allow add custom attributes inside form tag? Or I could remove it and upload the image using ajax and adding here a hidden input with the url for the user meta

SDavisMedia commented:
@rubengc for something that custom that so few people would need, I'd rather see you just override the entire template in your theme: http://docs.easydigitaldownloads.com/article/1216-moving-edd-templates-to-your-theme

There's an argument that the other hooks needed to be added anyway but I don't think it's a good idea to put one directly in the form tag on any of our forms in core. Let's go ahead and remove that and we should be clear to merge.

rubengc commented:
I removed edd_profile_editor_form_tag action

Form enctype attribute is not for theme, It is for plugin support, but not problem, I will upload the image using ajax :)

SDavisMedia commented:
@rubengc I understand it's not for your theme. I'm saying if you wanted to add an attribute to the form tag, override the entire template file from your theme (see the documentation I linked) and add the attribute there. Then you'd still have to the option to either use ajax or enctype attribute. We just don't want to put it in EDD core.

I'll get this reviewed again and hopefully merged. Thanks again!

cklosowski commented:
@SDavisMedia looks good on my end, just let me know when yo'ure done with your testing and i'll merge it away.

SDavisMedia commented:
@cklosowski I'm already good to go. Merge away, sir!

rubengc commented:
👍
